Aggregate service feedbacks and remove from Anonymous contacts table

binaryberry opened the pull request:
Service feedbacks are ratings that are requested of users at the end of a transaction - at the moment they represent 16 million rows out of the 17 million rows of anonymous contacts, and therefore slow down queries considerably, to the point they sometimes timeout. 

This removes service feedback ratings from the anonymous contacts table unless they have attached comments, and saves them to a separate table, the ArchivedServiceFeedback table. It also aggregates the count and adds rows to the feedex table so that users can see how many ratings of 1, 2, 3, 4, and 5 they received each day for each path.

Tested in integration and works with [this support PR](https://github.com/alphagov/support/pull/277).

Before:
![screen shot 2016-05-27 at 17 18 00](https://cloud.githubusercontent.com/assets/8225167/15614148/0b1f5df6-242f-11e6-8778-5bfc2374b80c.png)

After:
![screen shot 2016-06-06 at 11 52 10](https://cloud.githubusercontent.com/assets/8225167/15820465/7cd89638-2be1-11e6-96c9-8c079920366a.png)

After this, we will need to run the aggregation task on all existing rows, run Vacuum Analyze in the database in production, and double check the aggregation task gets run daily.

[Trello](https://trello.com/c/zhqhIKOs/377-reduce-duplication-of-quantitative-results-in-feedex-medium)

binaryberry commented on db/structure.sql:
```diff
@@ -271,41 +319,6 @@ CREATE INDEX index_anonymous_contacts_on_created_at ON anonymous_contacts USING
 
 
 --
--- Name: index_anonymous_contacts_on_created_at_and_path; Type: INDEX; Schema: public; Owner: -; Tablespace: 
```
@boffbowsh a bit puzzled as to why running the migration removed some indexes... is that normal?

boffbowsh replied:
This means that your database schema that it is created this structure from is missing these indexes. Perhaps they didn't get restored correctly from the integration backup during replication. Best to remove these lines from your commit.

binaryberry replied:
I did now! The diff is still a bit polluted by what I think is my editor's automated removal of spaces at the end of line. Hopefully it's no biggie but let me know if that was an issue.

boffbowsh replied:
Yeah, can you remove those changes from the diff please? Else next time the schema is regenerated it'll all just change again.

binaryberry replied:
Done now :-)

boffbowsh commented on lib/tasks/service_feedback_aggregation.rake:
```diff
@@ -0,0 +1,40 @@
+    p "Processing feedback from #{date_range.first} to #{date_range.last}"
+    date_range.each do |date|
+      aggregator = ServiceFeedbackAggregator.new
+      aggregator.run(date)
```
Might be good to print these dates to the console as they're run, and timestamped so whoever is running it can get an idea of how quickly it's running.

binaryberry replied:
Done now in bb0844e. For each date, outputting the date + duration of the task and removed total duration as it seemed superfluous.

boffbowsh commented on lib/tasks/service_feedback_aggregation.rake:
```diff
@@ -0,0 +1,40 @@
+
+  def aggregate(date_range)
+    start_time = Time.now
+    p "Processing feedback from #{date_range.first} to #{date_range.last}"
```
Should use `puts` rather than `p`.

binaryberry replied:
Thanks! addressed now.

boffbowsh commented on lib/tasks/service_feedback_aggregation.rake:
```diff
@@ -0,0 +1,40 @@
+    end_time = Time.now
+    p "Service Feedback has been aggregated and extracted to a different table"
+    p "duration = #{end_time - start_time}"
+  end
```
Might be good to run a Postgres [`VACUUM ANALYZE`](http://www.postgresql.org/docs/9.3/static/sql-vacuum.html) statement on `anonymous_contacts` at the end of the run. Once everything has been rolled up, we can schedule a `VACUUM FULL` for out of hours to fully reclaim all the space this operation is saving.

fofr commented on README.md:
```diff
@@ -20,3 +20,7 @@ to populate this in your development environment:
+
+## Data
+
+When [replicating data](https://github.gds/gds/development#8-import-production-data), by default the anonymous feedback is not imported. You will have to import it by going into development and running `./sync-postgresql.sh -r -s -d backups/YYYY-MM-DD postgresql-primary-1.backend.integration`, where the date corresponds to the date of your latest backup.
```
Good instructions! :100:

Davidslv commented on app/models/aggregated_service_feedback.rb:
```diff
@@ -0,0 +1,18 @@
+  end
+
+  def as_json(options = {})
+    attributes_to_serialise = [
```
you can use `%i`  

```
> %i(a b)
=> [:a, :b]
```

Not a blocker.

binaryberry replied:
Better indeed. I would prefer to keep it this way to conserve consistency with the other models ([long_form_contact](https://github.com/alphagov/support-api/blob/master/app/models/long_form_contact.rb#L10), [organisation](https://github.com/alphagov/support-api/blob/master/app/models/organisation.rb#L11), [problem report](https://github.com/alphagov/support-api/blob/master/app/models/problem_report.rb#L21) and [service feedback](https://github.com/alphagov/support-api/blob/master/app/models/service_feedback.rb#L10)). The styleguide says that "Consistency with this style guide is important. Consistency within a project is more important."

Davidslv commented on lib/service_feedback_aggregator.rb:
```diff
@@ -0,0 +1,71 @@
+class ServiceFeedbackAggregator
+
+  def run(date)
+    @date = date
```
This looks weird here.
You still have to initialize `ServiceFeedbackAggregator` so why not give it the date there?
Otherwise this feels like a class method.

binaryberry replied:
Good point @Davidslv, and I had initially written it like that. The reason I opted for this way was that in the rake task that uses this, we are going to process large amounts of service feedback, and if we only have to instantiate the ServiceFeedbackAggregator once instead of once per date, I think we'll save a bit of time? Not dead set on this though.

boffbowsh replied:
I think @Davidslv is right, the class should be initialised once per date, then have `run` called on it. This ensures that any state doesn't leak between runs. Instantiation of a PORO like this is relatively inexpensive.

binaryberry replied:
Gotcha! Thanks for the explanation. Amended in 4850364.

boffbowsh commented on lib/service_feedback_aggregator.rb:
```diff
@@ -0,0 +1,71 @@
+      FROM anonymous_contacts
+      WHERE type = 'ServiceFeedback'
+      AND created_at >= '#{@date}'
+      AND created_at < '(#{@date + 1.day})'
```
What are the parentheses for here?

binaryberry replied:
To test your eyes: you passed! Amended in relevant commits.

boffbowsh replied:
This one is still here

boffbowsh commented on lib/tasks/service_feedback_aggregation.rake:
```diff
@@ -0,0 +1,40 @@
+  task :date => :environment do
+    require 'service_feedback_aggregator'
+
+    puts "Please enter the date you would like aggregation to start from, in the format YYYY-MM-DD:"
```
Can we parameterise these for the rake task please, then we can run the script from within Jenkins.

binaryberry replied:
Sure, done in 4850364.

boffbowsh commented on spec/models/service_feedback_aggregator_spec.rb:
```diff
@@ -0,0 +1,100 @@
+require 'service_feedback_aggregator'
+
+describe ServiceFeedbackAggregator do
+  let(:date)  { Time.new(2013,2,10,10)  }
```
You've declared `date` helpers but there's a few places your'e still using the literal date in the code below

binaryberry replied:
Good point, done in 4850364.

boffbowsh commented on spec/models/service_feedback_aggregator_spec.rb:
```diff
@@ -0,0 +1,100 @@
+    let(:date) { Time.new(2013,2,11) }
+
+    it "copies service feedbacks to the archived service feedback table" do
+      expect(ArchivedServiceFeedback.count).to eq 0
```
You can use the expect change matcher for these: http://www.relishapp.com/rspec/rspec-expectations/v/3-4/docs/built-in-matchers/change-matcher

binaryberry replied:
Nice! Addressed in 4850364.

boffbowsh commented on spec/models/service_feedback_aggregator_spec.rb:
```diff
@@ -0,0 +1,100 @@
+
+    context "when there are two service feedbacks" do
+      context "with the same rating" do
+        it "there is one aggregated service feedback with a rating of 1" do
```
This doesn't read correctly. The example should read as english, including the `it`.
```ruby
it "creates an aggregated service feedback with a rating of 1" do
```

binaryberry replied:
Indeed! Amended now in 4850364.

boffbowsh commented on spec/models/service_feedback_aggregator_spec.rb:
```diff
@@ -0,0 +1,100 @@
+describe ServiceFeedbackAggregator do
+  let(:date)  { Time.new(2013,2,10,10)  }
+  let(:slug)  { "done/register-to-vote" }
+  let(:rating){ 1                       }
```
It's ok to not have braces align. Personally I think the whitespace here looks overly forced.

binaryberry replied:
oki dok! Amended in 4850364.

boffbowsh commented on spec/models/service_feedback_aggregator_spec.rb:
```diff
@@ -0,0 +1,100 @@
+    context "when there are two service feedbacks" do
+      context "with the same rating" do
+        it "there is one aggregated service feedback with a rating of 1" do
+          expect(AggregatedServiceFeedback.all.map(&:service_satisfaction_rating)).to eq [1]
```
You can use `pluck` here instead of `map`

binaryberry replied:
Done, thanks! 4850364

binaryberry commented:
@boffbowsh I believe I've addressed all comments, and I also added 60b2287 to avoid duplicate service feedback aggregates

boffbowsh commented on lib/service_feedback_aggregator.rb:
```diff
@@ -0,0 +1,88 @@
+  end
+
+  def run
+    return reason_for_not_running if reason_for_not_running
```
This doesn't get exposed anywhere

binaryberry replied:
Thanks! Added in 7ac6928.

gpeng commented on lib/service_feedback_aggregator.rb:
```diff
@@ -0,0 +1,93 @@
+  end
+
+  def run
+    return @reason_for_not_running if check_reason_for_not_running
```
This use the accessor `reason_for_not_running` rather than the variable `@reason_for_not_running`

binaryberry replied:
Thanks! Addressed now.

binaryberry commented:
@boffbowsh : I addressed the remaining parenthesis in the SQL query, and updated the daily rake task to only include yesterday's date, not today's. Should be good to go now, hopefully!
